packages/mds-ingest-service/repository/index.ts

```diff
@@ -14,14 +14,222 @@
+        .getRepository(DeviceEntity)
+        .createQueryBuilder()
+        .insert()
+        .values(events.map(DeviceDomainToEntityCreate.mapper()))
+        .returning('*')
+        .execute()
+      return entities.map(DeviceEntityToDomain.map)
+    } catch (error) {
+      throw RepositoryError(error)
+    }
+  }
+
+  public getEvents = async (params: GetVehicleEventsFilterParams): Promise<EventDomainModel[]> => {
+    const { connect } = this
+    const {
+      time_range: { start, end },
+      // geography_ids,
+      grouping_type = 'latest_per_vehicle',
+      event_types,
+      vehicle_states,
+      vehicle_types,
+      device_or_vehicle_id,
+      device_ids,
+      propulsion_types,
+      provider_ids
+    } = params
+    try {
+      const connection = await connect('ro')
+
+      const query = connection
+        .getRepository(EventEntity)
+        .createQueryBuilder('events')
+        .innerJoin(qb => qb.from(DeviceEntity, 'd'), 'devices', 'devices.device_id = events.device_id')
+        .leftJoinAndMapOne(
+          'events.telemetry',
+          TelemetryEntity,
+          'telemetry',
+          'telemetry.device_id = events.device_id AND telemetry.timestamp = events.telemetry_timestamp'
+        )
+
+      if (grouping_type === 'latest_per_vehicle') {
+        query.innerJoin(
+          qb => {
+            return qb
+              .select(
+                'device_id, id as event_id, RANK() OVER (PARTITION BY device_id ORDER BY timestamp DESC) AS rownum'
+              )
+              .from(EventEntity, 'e')
+              .where('timestamp >= :start AND timestamp <= :end', { start, end })
+          },
+          'last_device_event',
+          'last_device_event.event_id = events.id AND last_device_event.rownum = 1'
+        )
+      }
+
+      if (grouping_type === 'latest_per_trip') {
+        query.innerJoin(
+          qb => {
+            return qb
+              .select('trip_id, id as event_id, RANK() OVER (PARTITION BY trip_id ORDER BY timestamp DESC) AS rownum')
+              .from(EventEntity, 'e')
+              .where('timestamp >= :start AND timestamp <= :end', { start, end })
+          },
+          'last_trip_event',
+          'last_trip_event.event_id = events.id AND last_trip_event.rownum = 1'
+        )
+      }
+
+      if (event_types) {
+        query.andWhere('events.event_types && :event_types', { event_types })
+      }
+
+      if (propulsion_types) {
+        query.andWhere('devices.propulsion_types && :propulsion_types', { propulsion_types })
+      }
+
+      if (device_ids) {
+        query.andWhere('events.device_id = ANY(:device_ids)', { device_ids })
+      }
+
+      if (vehicle_types) {
+        query.andWhere('devices.vehicle_type = ANY(:vehicle_types)', { vehicle_types })
+      }
+
+      if (vehicle_states) {
+        query.andWhere('events.vehicle_state = ANY(:vehicle_states)', { vehicle_states })
+      }
+
+      if (device_or_vehicle_id) {
```

Comment:
I think it would be better to have separate `device_id` and `vehicle_id` filters in the repository (and maybe higher in the stack as well). In fact, there is already a `device_ids` filter above, but maybe I'm missing something. Also, does it make sense to use `LIKE ANY()` on an array of vehicles_id values?

Reply:
I'm just matching the API design that was requested from OA, so I haven't tried to factor it into anything more general purpose, until there's another thing that also wants to query on mds events.